packages/plugin-ext/doc/how-to-add-new-custom-plugin-api.md

```diff
@@ -0,0 +1,281 @@
+# How to add new custom plugin API
+
+As a Theia developer, you might want to make your app extensible by plugins in ways that are unique to your application.
+That will require API that goes beyond what's in the VS Code Extension API and the Theia plugin API.
+You can do that by implementing a Theia extension that creates and exposes an API object within the plugin host.
+The API object can be imported by your plugins and exposes one or more API namespaces.
+
+Depending on the plugin host we can either provide a frontend or backend plugin API:
+
+- In the backend plugin host that runs in the Node environment in a separate process, we adapt the module loading to return a custom API object instead of loading a module with a particular name.
+- In the frontend plugin host that runs in the browser environment via a web worker, we import the API scripts and put it in the global context.
+
+In this document we focus on the implementation of a custom backend plugin API.
+However, both APIs can be provided by implementing and binding an `ExtPluginApiProvider` which should be packaged as a Theia extension.
+
+## Declare your plugin API provider
+
+The plugin API provider is executed on the respective plugin host to add your custom API object and namespaces.
+
+Example Foo Plugin API provider:
+
+```typescript
+@injectable()
+export class FooExtPluginApiProvider implements ExtPluginApiProvider {
+    provideApi(): ExtPluginApi {
+        return {
+            frontendExtApi: {
+                initPath: '/path/to/foo/api/implementation.js',
+                initFunction: 'fooInitializationFunction',
+                initVariable: 'foo_global_variable'
+            },
+            backendInitPath: path.join(__dirname, 'foo-init')
+        };
+    }
+}
+```
+
+Register your Plugin API provider in a backend module:
+
+```typescript
+    bind(FooExtPluginApiProvider).toSelf().inSingletonScope();
+    bind(Symbol.for(ExtPluginApiProvider)).toService(FooExtPluginApiProvider);
+```
+
+## Define your API
+
+To ease the usage of your API, it should be developed as separate npm package that can be easily imported without any additional dependencies, cf, the VS Code API or the Theia Plugin API.
+
+Example `foo.d.ts`:
+
+```typescript
+declare module '@bar/foo' {
+    export namespace fooBar {
+        export function getFoo(): Promise<Foo>;
+    }
+}
+```
+
+## Implement your plugin API provider
+
+In our example, we aim to provide a new API object for the backend.
+Theia expects that the `backendInitPath` that we specified in our API provider is a function called `provideApi` that follows the `ExtPluginApiBackendInitializationFn` signature.
+
+Example `foo-init.ts`:
```

Comment:
```suggestion
Example `node/foo-init.ts`:
```

Reply:
I think it might make sense to include suitable parent directories for the sample files

Reply:
I love that idea, I added them as you suggested.